### Import medianprice_model

In [1]:
from find_medianprice_model import medianPriceModel
from median_price_model import MedianPriceModel
import flood_tool.geo
import warnings

warnings.filterwarnings("ignore")


The above block of code imports our model for predicting the median house price, which we will then use to predict the test data.

### Select best model

In [2]:
medianPriceModel().load_data().preprocessing().model_selection()

RandomForestRegressor()

In [3]:
type(medianPriceModel().load_data().preprocessing().model_selection())

sklearn.ensemble._forest.RandomForestRegressor

[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=5, regressor__n_estimators=3; total time=   0.1s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=5, regressor__n_estimators=20; total time=   0.6s
[CV] END regressor__bootstrap=True, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=5, regressor__n_estimators=11; total time=   0.1s
[CV] END regressor__bootstrap=True, regressor__max_depth=20, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=5, regressor__n_estimators=3; total time=   0.1s
[CV] END regressor__bootstrap=True, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=10, regressor__n_estimators=11; total tim

[CV] END regressor__bootstrap=True, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=10, regressor__n_estimators=13; total time=   0.2s
[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=10, regressor__n_estimators=15; total time=   0.3s
[CV] END regressor__bootstrap=False, regressor__max_depth=20, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=10, regressor__n_estimators=7; total time=   0.2s
[CV] END regressor__bootstrap=True, regressor__max_depth=20, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=10, regressor__n_estimators=15; total time=   0.2s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=5, regressor__n_estimators=13; total 

[CV] END regressor__bootstrap=True, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=5, regressor__n_estimators=20; total time=   0.4s
[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=5, regressor__n_estimators=5; total time=   0.1s
[CV] END regressor__bootstrap=True, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=2, regressor__n_estimators=13; total time=   0.2s
[CV] END regressor__bootstrap=False, regressor__max_depth=20, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=10, regressor__n_estimators=1; total time=   0.0s
[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=2, regressor__n_estimators=5; total time=

The output of the code block above shows that RandomForestRegressor gets the best results, with a smaller root_mean_squared_error. We use KNN regression as model_1 and random forest regression as model_2.

### Save best models 

In [3]:
model_1, params_1, score_1 = medianPriceModel().load_data().preprocessing().model_1()

832882.45


In [4]:
model_1

Pipeline(steps=[('preprocessing',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('ordinalencoder',
                                                  OrdinalEncoder(),
                                                  ['soilType',
                                                   'localAuthority']),
                                                 ('minmaxscaler',
                                                  MinMaxScaler(),
                                                  ['altitude'])])),
                ('regressor',
                 KNeighborsRegressor(n_neighbors=23, p=1, weights='distance'))])

**KNN model**: The code block above we derived the mean of the median of all house prices in the test set (832882.45), as well as the optimal parameters for the KNN model, the final root_mean_squared_error (**RMSE= -2832505.6118755457**).

In [5]:
params_1, score_1

({'regressor__n_neighbors': 23,
  'regressor__p': 1,
  'regressor__weights': 'distance'},
 -2832505.6118755457)

In [2]:
model_2, params_2, score_2 = medianPriceModel().load_data().preprocessing().model_2()

Fitting 2 folds for each of 100 candidates, totalling 200 fits
855485.0625


In [9]:
model_2

Pipeline(steps=[('preprocessing',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('ordinalencoder',
                                                  OrdinalEncoder(),
                                                  ['soilType',
                                                   'localAuthority']),
                                                 ('minmaxscaler',
                                                  MinMaxScaler(),
                                                  ['altitude'])])),
                ('regressor',
                 RandomForestRegressor(max_features='sqrt', min_samples_leaf=4,
                                       min_samples_split=10, n_estimators=3))])

[CV] END regressor__bootstrap=True, regressor__max_depth=None, regressor__max_features=auto, regressor__min_samples_leaf=4, regressor__min_samples_split=2, regressor__n_estimators=7; total time=   0.2s
[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=auto, regressor__min_samples_leaf=4, regressor__min_samples_split=5, regressor__n_estimators=13; total time=   0.2s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=10, regressor__n_estimators=17; total time=   0.2s
[CV] END regressor__bootstrap=True, regressor__max_depth=20, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=2, regressor__n_estimators=7; total time=   0.1s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=auto, regressor__min_samples_leaf=1, regressor__min_samples_split=5, regressor__n_estimators=5; total t

[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=auto, regressor__min_samples_leaf=2, regressor__min_samples_split=5, regressor__n_estimators=17; total time=   0.3s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=10, regressor__n_estimators=3; total time=   0.0s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=10, regressor__n_estimators=17; total time=   0.2s
[CV] END regressor__bootstrap=True, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=5, regressor__n_estimators=11; total time=   0.1s
[CV] END regressor__bootstrap=True, regressor__max_depth=10, regressor__max_features=auto, regressor__min_samples_leaf=2, regressor__min_samples_split=10, regressor__n_estimators=13; tot

[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=auto, regressor__min_samples_leaf=2, regressor__min_samples_split=5, regressor__n_estimators=17; total time=   0.3s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=auto, regressor__min_samples_leaf=1, regressor__min_samples_split=10, regressor__n_estimators=7; total time=   0.2s
[CV] END regressor__bootstrap=True, regressor__max_depth=None, regressor__max_features=auto, regressor__min_samples_leaf=4, regressor__min_samples_split=10, regressor__n_estimators=17; total time=   0.3s
[CV] END regressor__bootstrap=True, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=2, regressor__n_estimators=11; total time=   0.1s
[CV] END regressor__bootstrap=False, regressor__max_depth=20, regressor__max_features=sqrt, regressor__min_samples_leaf=1, regressor__min_samples_split=10, regressor__n_estimators=15; tot

[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=auto, regressor__min_samples_leaf=1, regressor__min_samples_split=2, regressor__n_estimators=17; total time=   0.3s
[CV] END regressor__bootstrap=False, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=10, regressor__n_estimators=3; total time=   0.0s
[CV] END regressor__bootstrap=False, regressor__max_depth=10, regressor__max_features=auto, regressor__min_samples_leaf=2, regressor__min_samples_split=5, regressor__n_estimators=3; total time=   0.1s
[CV] END regressor__bootstrap=True, regressor__max_depth=10, regressor__max_features=sqrt, regressor__min_samples_leaf=4, regressor__min_samples_split=2, regressor__n_estimators=17; total time=   0.1s
[CV] END regressor__bootstrap=True, regressor__max_depth=None, regressor__max_features=sqrt, regressor__min_samples_leaf=2, regressor__min_samples_split=5, regressor__n_estimators=11; total ti

In [7]:
params_2, score_2 

({'regressor__n_estimators': 3,
  'regressor__min_samples_split': 10,
  'regressor__min_samples_leaf': 4,
  'regressor__max_features': 'sqrt',
  'regressor__max_depth': None,
  'regressor__bootstrap': True},
 -3334808.5767527814)

**RandomForest model**: The code block above we derived the mean of the median of all house prices in the test set (888343.0625), as well as the optimal parameters for the KNN model, the final root_mean_squared_error (**RMSE=-3334808.5767527814**).

We found that the KNN model has a better RMSE value when predicting, so **the KNN model** is chosen as the final prediction model.

### Predict median price

In [52]:
data = pd.read_csv("resources/postcodes_sampled.csv")

median_price_model = MedianPriceModel(data)
trained_mp_model = median_price_model.train()

unlabelled_data = pd.read_csv("resources/postcodes_unlabelled.csv")
lat,lon=geo.get_gps_lat_long_from_easting_northing(list(unlabelled_data["easting"]),list(unlabelled_data["northing"]))
unlabelled_data["latitude"] = lat 
unlabelled_data["longitude"] = lon
unlabelled_data = unlabelled_data[["soilType","localAuthority","altitude","latitude","longitude"]]
y_pred = median_price_model.predict(trained_mp_model, unlabelled_data)
print(y_pred)

[567665.44015911 614543.01116181 695300.         ... 301870.58082426
 555390.2821279  443825.26350896]


In [47]:
prediction = pd.DataFrame(y_pred)
prediction

0
0     567665.440159
1     614543.011162
2     695300.000000
3     486309.406243
4     434625.845110
...             ...
4995  528861.033504
4996  551708.081397
4997  301870.580824
4998  555390.282128
4999  443825.263509

[5000 rows x 1 columns]

In [48]:
# df = pd.concat([df,prediction], axis=1,join="inner")
unlabelled_data['prediction'] = y_pred
unlabelled_data

soilType     localAuthority  altitude   latitude  longitude  \
0     Unsurveyed/Urban  Brighton and Hove        60  50.834677  -0.148139   
1     Unsurveyed/Urban              Lewes        10  50.879937   0.017610   
2     Unsurveyed/Urban            Wealden       160  51.041906   0.170125   
3     Unsurveyed/Urban  Brighton and Hove        60  50.845393  -0.189664   
4     Unsurveyed/Urban             Bexley        60  51.458048   0.101839   
...                ...                ...       ...        ...        ...   
4995  Unsurveyed/Urban           Haringey        10  51.598652  -0.055595   
4996  Unsurveyed/Urban            Hackney        10  51.543341  -0.046618   
4997  Unsurveyed/Urban         Canterbury        10  51.370422   1.120780   
4998  Unsurveyed/Urban    Epsom and Ewell        40  51.345942  -0.252416   
4999  Unsurveyed/Urban         Spelthorne        10  51.400489  -0.461228   

         prediction  
0     567665.440159  
1     614543.011162  
2     695300.000000  
3     486309.406243  
4     434625.845110  
...             ...  
4995  528861.033504  
4996  551708.081397  
4997  301870.580824  
4998  555390.282128  
4999  443825.263509  

[5000 rows x 6 columns]

In [49]:
# save DataFrame as .csv file
outputpath = 'predict_result_medianprice.csv'
unlabelled_data.to_csv(outputpath,sep=',',index=False,header=True)